src/blob.jl

```diff
@@ -63,36 +63,40 @@ The number of bytes needed to allocate `T` itself.
 
 Defaults to `sizeof(T)`.
 """
-@generated function self_size(::Type{T}) where T
-    @assert isconcretetype(T)
+Base.@assume_effects :foldable function self_size(::Type{T}) where T
+    # This function is marked :total to encourage constant folding for this types-only
+    # static computation.
     if isempty(fieldnames(T))
-        quote
-            $(Expr(:meta, :inline))
-            $(sizeof(T))
-        end
+        sizeof(T)
     else
-        quote
-            $(Expr(:meta, :inline))
-            $(+(0, @splice i in 1:length(fieldnames(T)) begin
-                self_size(fieldtype(T, i))
-            end))
-        end
+        # Recursion is the fastest way to compile this, confirmed with benchmarks.
+        # Alternative considered: +(Iterators.map(self_size, fieldtypes(T))...)
+        # ~0.5ms for 5 fields, vs ~5ms for unrolling via splatting the fields.
+        # ~3ms for 20 fields, vs ~6ms for splatting.
+        # Note that splatting gives up after ~30 fields, whereas recursion remains robust.
+        _sum_field_sizes(T)
     end
 end
+Base.@assume_effects :foldable _sum_field_sizes(::Type{T}) where {T} =
+    _sum_field_sizes(T, Val(fieldcount(T)))
+Base.@assume_effects :foldable _sum_field_sizes(::Type, ::Val{0}) = 0
+Base.@assume_effects :foldable function _sum_field_sizes(::Type{T}, ::Val{i}) where {T,i}
+    return self_size(fieldtype(T, i)) + _sum_field_sizes(T, Val(i-1))
+end
 
-function blob_offset(::Type{T}, i::Int) where {T}
-    +(0, @splice j in 1:(i-1) begin
-        self_size(fieldtype(T, j))
-    end)
+# Recursion scales better than splatting for large numbers of fields.
+Base.@assume_effects :foldable @inline function blob_offset(::Type{T}, i::Int) where {T}
+    return _blob_offset(T, i-1)
+end
+Base.@assume_effects :foldable @inline function _blob_offset(::Type{T}, i::Int) where {T}
+    i <= 0 && return 0
+    return _blob_offset(T, i-1) + self_size(fieldtype(T, i))
 end
 
-@generated function Base.getindex(blob::Blob{T}, ::Type{Val{field}}) where {T, field}
+@inline function Base.getindex(blob::Blob{T}, ::Type{Val{field}}) where {T, field}
     i = findfirst(isequal(field), fieldnames(T))
-    @assert i != nothing "$T has no field $field"
-    quote
-        $(Expr(:meta, :inline))
-        Blob{$(fieldtype(T, i))}(blob + $(blob_offset(T, i)))
-    end
+    @assert i !== nothing "$T has no field $field"
+    Blob{fieldtype(T, i)}(blob + (blob_offset(T, i)))
```

Comment:
The `+` creates a `Blob{T}` that we then cast to `Blob{fieldtype(T, i)}`. Wouldn't it be better to create the right type from the beginning? (I think the +/- operators don't make much sense)

Reply:
That sounds reasonable to me. Again, i just did a blind transformation on what was here... 🤔 

I _think_ the + operators are adding bytes, in which case you could do it either way? But yes i agree this is confusing